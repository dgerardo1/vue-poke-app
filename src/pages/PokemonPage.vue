<template>

  <div class="container">

    <div v-if="!pokemon">
      <img
        src="https://fontmeme.com/permalink/211223/705fb7315efa6567446cf7f9c65b0ce1.png"
        alt="fuente-pokemon"
        border="0">
    </div>
    
    <div v-else>
    
        <img
          src="https://fontmeme.com/permalink/211208/abb56b455692b0a09af69a5d8d3135d3.png"
          alt="fuente-pokemon"
          border="0">
    
      <PokemonPicture
        :pokemonId="pokemon.id"
        :showPokemon="showPokemon"/>
      
      <PokemonOptions
        :pokemons="pokemonArr"
        @selection="checkAnswer"/>

       <template v-if="showAnswer">
          <h2>{{ message }}</h2>
          <button @click="newGame">
            Nuevo Juego
          </button>
       </template>
    
    </div>

  </div>

</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonOptions'

// console.log( getPokemonOptions() )

export default {
  components: { 
    PokemonPicture, 
    PokemonOptions 
  },
  data(){
    return{
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async mixPokemonArray(){

      this.pokemonArr = await getPokemonOptions()
      // console.log(this.pokemonArr)

      const rndInt = Math.floor( Math.random() * 4 )
      // console.log(rndInt)

      this.pokemon = this.pokemonArr[ rndInt ]

    },
    checkAnswer( selectedId ){

      // console.log('Pokemon page llamado', pokemonId )
      this.showPokemon = true
      this.showAnswer = true

      if ( selectedId === this.pokemon.id ) {
        this.message = `¡Correcto! Es ${ this.pokemon.name[0].toUpperCase() + this.pokemon.name.substring(1) }`
      } else {
        this.message = `¡Ups! Era ${ this.pokemon.name[0].toUpperCase() + this.pokemon.name.substring(1) }`        
      }

    },
    newGame(){
      this.showPokemon = false
      this.showAnswer  = false
      this.pokemonArr  = []
      this.pokemon     = null
      this.mixPokemonArray()
    }
  },
  mounted() {
    this.mixPokemonArray()
  }

}
</script>

<style scoped>
  .container {
    max-width: 720px;
    margin: 0 auto;
  }

  img {
    width: 80%;
  }

  h2 {
    color: white !important;
    margin-top: 0;
  }

  button {
    padding: 6px 12px;
    background-color: #ffcb05;
    border-radius: 6px;
    font-size: 16px;
    cursor: pointer;
    color: rgba(229,73,70,1);
    border: 2px solid rgba(42, 117, 187, 1);
  }

</style>