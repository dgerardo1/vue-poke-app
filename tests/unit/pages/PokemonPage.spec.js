import { shallowMount, mount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import { pokemons } from '../mocks/pokemons.mock';

describe('PokemonPage Component', () => {

    let wrapper

    beforeEach( () => {        
        wrapper = shallowMount( PokemonPage )
    })

    test('debe de hacer match con el snapshot', () => {
        
        expect( wrapper.html() ).toMatchSnapshot()
    });

    test('debe de llamar el mixPokemonArray al montar', () => {

        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )
        shallowMount( PokemonPage )

        expect( mixPokemonArraySpy ).toHaveBeenCalled()
        
    });

    test('debe de hacer match con el snapshot cuando cargan los pokemons', () => {
        
        const wrapper = shallowMount( PokemonPage, {
            data(){
                return{
                  pokemonArr: pokemons,
                  pokemon: pokemons[0],
                  showPokemon: false,
                  showAnswer: false,
                  message: ''
                }
            }
        } )

        expect( wrapper.html() ).toMatchSnapshot()
    });

    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {

        const wrapper = shallowMount( PokemonPage, {
            data(){
                return{
                  pokemonArr: pokemons,
                  pokemon: pokemons[0],
                  showPokemon: false,
                  showAnswer: false,
                  message: ''
                }
            }
        } )

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        // PokemonPicture debe existir        
        expect( picture.exists() ).toBe( true )

        // PokemonPicture debe existir
        expect( options.exists() ).toBe( true )

        // PokemonPicture attribute pokemonid === 1        
        expect( picture.attributes('pokemonid') ).toBe( '1' )

        // PokemonOptions attribute pokemons toBe true
        expect( options.attributes('pokemons') ).toBeTruthy()
        
    });

    test('pruebas con checkAnswer', async() => {
        
        const wrapper = shallowMount( PokemonPage, {
            data(){
                return{
                  pokemonArr: pokemons,
                  pokemon: pokemons[0],
                  showPokemon: false,
                  showAnswer: false,
                  message: ''
                }
            }
        } )

        await wrapper.vm.checkAnswer(1)

        expect( wrapper.find('h2').exists() ).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBe( true )
        expect( wrapper.find('h2').text() ).toBe( `¡Correcto! Es ${ pokemons[0].name[0].toUpperCase() + pokemons[0].name.substring(1) }` )

        await wrapper.vm.checkAnswer(10)
        // console.log(wrapper.vm.message )
        expect( wrapper.vm.message ).toBe( `¡Ups! Era ${ pokemons[0].name[0].toUpperCase() + pokemons[0].name.substring(1) }` )

    });
    
});