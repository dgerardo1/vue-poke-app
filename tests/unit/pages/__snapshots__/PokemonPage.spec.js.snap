// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot 1`] = `
<div class="container">
  <div><img src="https://fontmeme.com/permalink/211223/705fb7315efa6567446cf7f9c65b0ce1.png" alt="fuente-pokemon" border="0"></div>
</div>
`;

exports[`PokemonPage Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div class="container">
  <div><img src="https://fontmeme.com/permalink/211208/abb56b455692b0a09af69a5d8d3135d3.png" alt="fuente-pokemon" border="0">
    <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
    <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    <!--v-if-->
  </div>
</div>
`;
